### PR TITLE
Enable restore for ref and src projects in libraries

### LINKED ISSUE #29953: Should test restore be part of build or build-tests?
With the test project restore work test projects need to be individually restored before building.

Today this was added to the overall build.cmd, but this is not a necessary step for build and lengthens the time it takes for a build to complete.  On my machine an incremental build with no work to do, this is taking 2 minutes for a no-op restore of the test projects on my notebook.

Can we consider moving this to the build-tests phase?  So long as an un-restored test project still errors correctly when trying to build without restoring it seems like a reasonable optimization.  

Thoughts @stephentoub @safern @wtgodbe @ViktorHofer?  Anyone else have numbers here?

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/runtime/issues/29953
Contributes towards https://github.com/dotnet/runtime/issues/31844

Upfront restore is now taking far too long, presumably because of project evaluation or multiple globs. I'm considering making restore implicit (during the build of the refs, sources and tests) so that the initial restore only handles the depproj files. Under a switch the restore target would still be able to upfront restore everything (CI mode).

Later when we removed all depproj files and sequenced _everything_ into the build, we can remove the `-restore` action which is passed to the Arcade's Build.proj so that the `-restore` action is only invoked on demand (CI or before going offline, ie airplane).

### Requires
- [x] https://github.com/dotnet/arcade/pull/4989
- [x] https://github.com/dotnet/arcade/pull/4992
- [x] https://github.com/dotnet/arcade/pull/4993
- [x] https://github.com/dotnet/arcade/pull/4996
- [ ] https://github.com/dotnet/sdk/pull/10794

cc @Anipik @joperezr @safern

### CODE REVIEW

File: eng/Analyzers.props [hunk 1]
```diff
@@ -1,6 +1,8 @@
 <Project>
   <PropertyGroup>
     <CodeAnalysisRuleset>$(MSBuildThisFileDirectory)CodeAnalysis.ruleset</CodeAnalysisRuleset>
+    <!-- Disable using analyzers in sourcebuild -->
+    <EnableAnalyzers Condition="'$(DotNetBuildFromSource)' == 'true'">false</EnableAnalyzers>
```
Comment (joperezr) [inline eng/Analyzers.props]: This doesn't seem to be related to this change right? perhaps we should separate it on a different PR?

Comment (ViktorHofer) [reply eng/Analyzers.props]: This is related, as before src and ref assemblies were going through the manual target that @safern wrote which adds the Analyzers to the csc task. Now they go through the same path as test assemblies and other libs like CoreLib. I moved this out from Tools.props and applied the condition here for sourcebuild.

File: eng/Tools.props [hunk 2]
```diff
@@ -1,45 +1,15 @@
 <Project>
-  <PropertyGroup>
-    <MSBuildTreatWarningsAsErrors>false</MSBuildTreatWarningsAsErrors>
```
Comment (joperezr) [inline eng/Tools.props]: We probably still want the TreatWarningsAsErrors right? In order to catch errors where one of the packages you were trying to restore wasn't available so you got a different one instead.

Comment (ViktorHofer) [reply eng/Tools.props]: Right but this was set to false before

File: eng/Tools.props [hunk 3]
```diff
@@ -1,45 +1,15 @@
...
-
-  <!-- excluded from offline portion of source build -->
-  <ItemGroup Condition="'$(DotNetBuildOffline)' != 'true'">
-    <!-- arcade -->
-    <PackageReference Include="Microsoft.DotNet.GenAPI" Version="$(MicrosoftDotNetGenApiVersion)" />
   </ItemGroup>
 
   <!-- excluded from source build -->
   <ItemGroup Condition="'$(DotNetBuildFromSource)' != 'true'">
-    <PackageReference Include="Microsoft.DotNet.ApiCompat" Version="$(MicrosoftDotNetApiCompatVersion)" />
```
Comment (joperezr) [inline eng/Tools.props]: Where will we be getting all of these references now? I haven't found where all of these tools reference went, same for the compilers toolset and Genfacades.

Comment (joperezr) [reply eng/Tools.props]: Oh nvm, I just found them 😄

Comment (ViktorHofer) [reply eng/Tools.props]: https://github.com/dotnet/runtime/pull/33242/files#diff-d3bedba39ce6da3e37ca0db83d7ef035R159-R164, and for the compiler https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Compiler.props#L5.

File: eng/Tools.props (hunk 2 above)
Comment (joperezr) [reply eng/Tools.props]: Ohh right! haha my bad.

File: src/libraries/ref.proj [hunk 4]
```diff
@@ -0,0 +1,36 @@
+<Project InitialTargets="IncludeProjectReferences">
```
Comment (joperezr) [inline src/libraries/ref.proj]: NIT: Looks like git is not considering this and ref.builds to have been the same file, which sometimes breaks the ability to check history on the file or to blame a line. When making these renames, can we try using `git mv oldFileName newFileName` so that history is preserved?

Comment (joperezr) [reply src/libraries/ref.proj]: It also makes it easier to review 😉

File: src/libraries/ref.proj [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  
+  <Target Name="Build" DependsOnTargets="BuildAllProjects" />
+  <Target Name="BuildAllProjects">
+    <MSBuild Targets="Build"
+             Projects="@(Project)"
+             Properties="BuildAllProjects=true;%(Project.AdditionalProperties)"
+             BuildInParallel="true"
+             ContinueOnError="ErrorAndStop" />
+
+    <!-- Given we ErrorAndContinue we need to propagate the error if the overall task failed -->
```
Comment (joperezr) [inline src/libraries/ref.proj]: Looks from above that now we have it as ErrorAndStop, so we probably don't need this last Error task any more?

File: src/libraries/src.proj [hunk 6]
```diff
@@ -17,5 +21,23 @@
...
-  <Import Project="$(MSBuildThisFileDirectory)dir.traversal.targets" />
+  <Target Name="IncludeProjectReferences">
+    <ItemGroup>
+      <ProjectReference Include="@(Project)" />
+    </ItemGroup>
+  </Target>
+
+  <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.targets" />
+  
+  <Target Name="Build" DependsOnTargets="BuildAllProjects" />
```
Comment (joperezr) [inline src/libraries/src.proj]: It might be worth it to add a comment above Build target to say that ordering matters and that we are intentionally defining this after importing sdk.targets to override that value, that way if people in the future edit this file will keep that in mind.

File: src/libraries/src.proj [hunk 7]
```diff
@@ -1,5 +1,9 @@
-<Project DefaultTargets="Build" TreatAsLocalProperty="AdditionalBuildTargetFrameworks">
-  <Import Project="$(MSBuildThisFileDirectory)Directory.Build.props" />
+<Project InitialTargets="IncludeProjectReferences">
+  <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" />
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp5.0</TargetFramework>
```
Comment (joperezr) [inline src/libraries/src.proj]: $(NetCoreAppCurrent) instead?

File: src/libraries/ref.proj (hunk 4 above)
Comment (ViktorHofer) [reply src/libraries/ref.proj]: just did that, thanks

File: src/libraries/ref.proj (hunk 5 above)
Comment (ViktorHofer) [reply src/libraries/ref.proj]: I believe in the past I copied that block from dir.traversal.targets. Let's try it out.

Comment (ViktorHofer) [reply src/libraries/ref.proj]: Oh I see, for testing it is ErrorAndContinue which makes sense.

File: src/libraries/src.proj (hunk 7 above)
Comment (ViktorHofer) [reply src/libraries/src.proj]: This one is problematic. I believe we need to set a TargetFramework here so that NuGet doesn't complain. But I'm not sure if this TargetFramework has any impact on the chosen projects. I don't know what the right way is here, maybe setting it to BuildTargetFramework? cc @Anipik 

This needs to be solved for ref.proj, src.proj and tests.proj.

File: src/libraries/Directory.Build.props [hunk 8]
```diff
@@ -351,6 +352,12 @@
   <!-- Import it at the end of the props file to override the OutputPath for reference assemblies and use common directory props -->
   <Import Project="$(RepositoryEngineeringDir)referenceAssemblies.props" />
 
+  <PropertyGroup>
+    <EnableProjectRestore Condition="'$(IsSourceProject)' == 'true' or '$(IsReferenceAssembly)' == 'true' or '$(IsTestProject)' == 'true'">true</EnableProjectRestore>
+  </PropertyGroup>
+
+  <Import Condition="'$(EnableProjectRestore)' != 'true'" Project="$(RepositoryEngineeringDir)restore\repoRestore.props" />
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/libraries/Directory.Build.props]: - depproj (which doesn't really count, as we never invoke the restore target as entrypoints on them)
- pkgproj

File: src/libraries/src.proj [hunk 9]
```diff
@@ -17,5 +21,21 @@
     <Project Include="$(DirectoryToBuild)\src\*.*proj" />
   </ItemGroup>
 
-  <Import Project="$(MSBuildThisFileDirectory)dir.traversal.targets" />
+  <Target Name="IncludeProjectReferences">
+    <ItemGroup>
+      <ProjectReference Include="@(Project)" />
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/libraries/src.proj]: Before with the old configuration system and the FilterProjects target, this was necessary (in tests.proj).

File: src/libraries/build.proj [hunk 10]
```diff
@@ -19,10 +19,12 @@
...
     </ItemGroup>
-    
-    <ItemGroup Condition="'$(DotNetBuildFromSource)' != 'true'">
-      <_RestoreProjects Include="$(MSBuildThisFileDirectory)tests.proj" Build="false" />
+
+    <ItemGroup>
+      <_RestoreProjects Include="$(MSBuildThisFileDirectory)dirs.proj" Build="false" />
+      <_RestoreProjects Include="$(MSBuildThisFileDirectory)tests.proj" Condition="'$(DotNetBuildFromSource)' != 'true'" Build="false" />
     </ItemGroup>
+
```
Comment (Anipik) [inline src/libraries/build.proj]: nit : extra line

File: src/libraries/dirs.proj [hunk 11]
```diff
@@ -1,26 +1,37 @@
...
+  <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" />
 
-  <!-- required to build the projects in their specified order -->
   <PropertyGroup>
-    <SerializeProjects>true</SerializeProjects>
+    <TargetFramework>$(BuildTargetFramework)</TargetFramework>
   </PropertyGroup>
 
   <ItemGroup Condition="'$(DirectoryToBuild)' == ''">
-    <Project Include="ref.builds" />
```
Comment (Anipik) [inline src/libraries/dirs.proj]: just for curiosity, why are we converting .build to projs ?

File: src/libraries/dirs.proj [hunk 12]
```diff
@@ -1,26 +1,37 @@
 <Project>
```
Comment (Anipik) [inline src/libraries/dirs.proj]: isnt adding ```<Project Sdk="Microsoft.NET.Sdk">``` will import both the sdk.props and sdk.targets ?

File: src/libraries/dirs.proj (hunk 11 above)
Comment (joperezr) [reply src/libraries/dirs.proj]: Yes, but he wants to control the exact time where the Sdk.targets get loaded, so he can then override the Build target. Check bellow where he actually added a comment about this.

Comment (ViktorHofer) [reply src/libraries/dirs.proj]: For Intellisense to work by default. .*proj is also restorable by nuget.

**Comment by ViktorHofer** [comment]
@ericstj this is ready. PTAL

File: src/libraries/Directory.Build.props (hunk 8 above)
[the remark this answers is not in the document]
Comment (ericstj) [reply src/libraries/Directory.Build.props]: I see, we should try to clean up this mode, even if it means doing something slightly different for these.  I was hoping to completely eliminate this hack.

File: src/libraries/dirs.proj [hunk 13]
```diff
@@ -1,26 +1,37 @@
...
   <ItemGroup Condition="'$(DirectoryToBuild)' == ''">
-    <Project Include="ref.builds" />
-    <Project Include="shims\generated\*.csproj" />
-    <Project Include="Native\native-binplace.proj" />
-    <Project Include="src.builds" />
-    <Project Include="shims\manual\*.csproj" />
-    <Project Include="shims\ApiCompat.proj" Condition="'$(DotNetBuildFromSource)' != 'true' and '$(BuildTargetFramework)' == '$(NetCoreAppCurrent)'" />
-    <Project Include="pretest.proj" Condition="'$(DotNetBuildFromSource)' != 'true' and '$(BuildAllConfigurations)' != 'true'" />
+    <ProjectReference Include="ref.proj" />
+    <ProjectReference Include="shims\generated\*.csproj" />
```
Comment (ericstj) [inline src/libraries/dirs.proj]: Can you move this into ref.proj and build as another target after the main build?  That would make ref.proj more useful, even though it does make it slightly more complicated.

File: tools-local/tasks/installer.tasks/installer.tasks.csproj [hunk 14]
```diff
@@ -20,6 +20,14 @@
     <PackageReference Include="System.Reflection.Metadata" Version="1.7.0" />
   </ItemGroup>
 
+  <!--
+    Restore the packaging package here to use it during the repo restore phase in 
+    harvestPackages.targets which is hooked into Tools.props.
```
Comment (ericstj) [inline tools-local/tasks/installer.tasks/installer.tasks.csproj]: Interesting hack.  I don't suppose there was a cleaner way to get this restored earlier?

**Review by ericstj — APPROVED** [review]

File: tools-local/tasks/installer.tasks/installer.tasks.csproj (hunk 14 above)
Comment (ViktorHofer) [reply tools-local/tasks/installer.tasks/installer.tasks.csproj]: Unfortunately not unless we bring the necessary task into the repo or sequence the harvesting package download out of the repo restore phase.

Comment (ericstj) [reply tools-local/tasks/installer.tasks/installer.tasks.csproj]: I see.  Do our Proj files support restore?  If so we could have repo restore download the package and packages.proj do the PackageDownloads.

File: src/libraries/dirs.proj (hunk 13 above)
Comment (ViktorHofer) [reply src/libraries/dirs.proj]: If you don't feel strongly about this I would prefer to not do so as we will delete these files (ref.proj, src.proj, tests.proj, dirs.proj, build.proj, package.builds) anyway as we want to define subsets and add the projects as ProjectToBuild arcade extension point items.

File: tools-local/tasks/installer.tasks/installer.tasks.csproj (hunk 14 above)
Comment (ViktorHofer) [reply tools-local/tasks/installer.tasks/installer.tasks.csproj]: That's a great idea. Yeah *proj files are restorable by default since a nuget update (I believe last year).

File: src/libraries/dirs.proj (hunk 13 above)
Comment (ericstj) [reply src/libraries/dirs.proj]: Sure, don't do anything that's throw away, but understand that dirs.proj is representing key sequencing dependencies, so I suspect these things will need to continue to be grouped somehow.  It'd be nice if they were grouped in a way that permitted folks to perform the common task of "build all inputs needed for building src projects" which is refs and shims.

Comment (ViktorHofer) [reply src/libraries/dirs.proj]: I was already in that file so I did it anyway :)

Yeah that's possible with the subsets. We can follow-up on that later in a few weeks.

**Comment by joperezr** [comment]
image


@ViktorHofer 😆 🤣

**Comment by jaredpar** [comment]
Oh man I laughed so hard when I read this it scared the dog

image

**Comment by ViktorHofer** [comment]
Failure is https://github.com/dotnet/runtime/issues/13295
